### PR TITLE
[#6789] Allow ENS names to be used for ethereum events

### LINKED ISSUE #6789: Allow ENS names to be used for ethereum events
### User Story

As a developer, I want to use ENS name instead of addresses in ethereum events.

### Description

*Type*: Feature

`ethereum/call` and `ethereum/send-transaction` should be able to resolve contracts from ENS name. An extra `ens` argument could be introduced.

### PR DESCRIPTION
**Pull request by tbenr**
Fixes #6789

### CODE REVIEW

File: src/status_im/extensions/ethereum.cljs [hunk 1]
```diff
@@ -116,16 +118,24 @@
...
\ No newline at end of file
+ (fn [{db :db} [_ _ {:keys [fromBlock toBlock address topics blockhash on-result] :as arguments}]]
+   (if (ens/is-valid-eth-name? address)
+     (let [{:keys [web3 network]} db
+           network-info (get-in db [:account/account :networks network])
+           chain (ethereum/network->chain-keyword network-info)
+           registry (get ens/ens-registries chain)]
+       (do
+         (print "network: " network "\nchain: " chain)
+         (ens/resolver web3 registry address #(re-frame/dispatch [:extensions/ethereum-logs nil [nil nil (assoc arguments :address %)]]))))
```
Comment (jeluard) [inline src/status_im/extensions/ethereum.cljs]: Can a avoid an extra dispatch and just provide a reference to a function?  The block in the else clause would be this function.

Comment (tbenr) [reply src/status_im/extensions/ethereum.cljs]: Yes, you're right. Btw the PR was just a placeholder for the gitcoin.

**Comment by tbenr** [comment]
what i'd like to do is a general method to extract addresses from the argument\s, resolve all addresses found, and then call to web3 using resolved addresse(s).
So it can be applied to all eth events dealing with addresses.
I need to study how to chain multiple address resolution (asynchronous calls with cb)

**Comment by jeluard** [comment]
@tbenr That's probably doable with a higher order fn (having the original fn as argument). This function could detect that `address` is passed as an argument then resolve it via ENS.

**Comment by tbenr** [comment]
@jeluard yes, but the complex part for me is to resolve multiple ens and then, when all done, call the fn.
for instance, for the `ethereum-logs` I can have a vector of addresses in `:address` parameter and also multiple addresses in the topic definition. 

let's say I add a `multiresolver` in ens that accepts a set of ens `#{"test.eth"}` and returns a map `{:test.eth "0x0000"}` 

so the events are responsible to:

1. extract ens from their arguments and generate the set of ens
2. call the `multiresolver`
3. recreate arguments by substituing ens with resolved addresses using the received map

**Comment by jeluard** [comment]
@tbenr Oh right.. Maybe as a first step ENS names could be resolved for the `address` field only?

**Comment by tbenr** [comment]
@jeluard that will be quite straight forward :-)
ok i'll do the simplified version than I can think on a more complete version on a separate PR

**Comment by tbenr** [comment]
@jeluard this is just a test commit, not tested. But, instead of implementing resolution for every event interface, what about exposing a dedicated event for ens resolution?
i see a couple of benefits be:
- simpler implementation (you don't need to care about where addresses are spreaded along the parameters)
- extension developer is "forced" to resolve  store resolved addresses in the state. (so they suppose to resolve addresses once)

what do you think?

**Comment by jeluard** [comment]
@tbenr  Actually I think both are useful ! I am fine with your current approach (makes it trivial to switch to ENS for those existing calls).

Now it would be useful too to expose ENS resolution as an event.

**Comment by tbenr** [comment]
@jeluard ok! are there some sample extensions to test the other ethereum events?

**Comment by jeluard** [comment]
@tbenr Not on the shelf but kyber has a contract registered using ENS so the following shoudl work:

```clojure
[ethereum/call {:to        "kybernetwork.eth" 
                   :method    "getExpectedRate(address,address,uint256)"
                   :params    ["0x744d70fdbe2ba4cf95131626614a1763df805b9e" "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee" 1000000000000000] 
                   :outputs   ["uint256" "uint256"] 
                   :on-result [on-src-picker-change-result {:selection selection}]}]
```

**Comment by tbenr** [comment]
@jeluard i added also a new event for the resolution (check if you like the name ;))

btw, tried to create a macro to simplify the ens resolution for every event.. something like:

```
(defmacro resolve-ens-and-call
 [ens-holder call]
 `(fn [{db :db} [_ _ {:keys [~ens-holder] :as arguments}]]
   (if (ens/is-valid-eth-name? ~ens-holder)
     (let [{:keys [web3 network]} db
           network-info (get-in db [:account/account :networks network])
           chain (ethereum/network->chain-keyword network-info)
           registry (get ens/ens-registries chain)]
       (do
         (print "\n\nnetwork: " network "\nchain: " chain)
         (ens/resolver web3 registry ens-holder #(~call (assoc arguments ~ens-holder %)))))
    (~call arguments))))

(handlers/register-handler-fx
 :extensions/ethereum-logs
 (resolve-ens-and-call :address call-get-logs))
```

but there's something basically wrong: i'm not able to manage macros yet :)

**Comment by jeluard** [comment]
@tbenr Yes the ENS event name sounds right :)

Did you try to create a function to wrap the call?

Something like:

```clojure
(defn wrap-with-resolution [f ...]
   (if (ens/is-valid-eth-name? to)
                          :data (when (and method params) (abi-spec/encode method params)))	     (let [{:keys [web3 network]} db
         transaction (prepare-extension-transaction tx-object (:contacts/contacts db) on-result)]	           network-info (get-in db [:account/account :networks network])
     (models.wallet/open-modal-wallet-for-transaction db transaction tx-object))))	           chain (ethereum/network->chain-keyword network-info)
           registry (get ens/ens-registries chain)]
       (ens/get-addr web3 registry to #(execute-send-transaction db (assoc db arguments :to %))))
     (f arguments))))

;; Then call it later on
(wrap-with-resolution #(execute-send-transaction db %) ...)
```

**Comment by tbenr** [comment]
I didn't realize macro was not needed to do that.
PR is finished.

**Review by jeluard — APPROVED** [review]

**Review by goranjovic — APPROVED** [review]

File: src/status_im/extensions/ethereum.cljs [hunk 2]
```diff
@@ -27,6 +29,21 @@
  (fn [{db :db} [_ on-result message]]
    (when on-result {:dispatch (on-result {:error message :result nil})})))
 
+(defn- wrap-with-resolution [db arguments address-keyword f]
```
Comment (flexsurfer) [inline src/status_im/extensions/ethereum.cljs:32]: what is `f` ?

**Review by flexsurfer — APPROVED** [review]

File: src/status_im/extensions/ethereum.cljs (hunk 2 above)
Comment (jeluard) [reply src/status_im/extensions/ethereum.cljs:32]: It's the function wrapped that will be called. Any better name?

**Comment by jeluard** [comment]
@tbenr Actually do you mind change `ethereum/resolve-ens` to `ethereum.ens/resolve` ? Sorry for the last minute change.

File: src/status_im/extensions/ethereum.cljs (hunk 2 above)
Comment (tbenr) [reply src/status_im/extensions/ethereum.cljs:32]: `cb`? `callback`? `eventfunction`?

**Pull request merged**